substrate/client/network/sync/src/strategy.rs

```diff
@@ -0,0 +1,484 @@
+		match self {
+			SyncingStrategy::WarpSyncStrategy(strategy) =>
+				strategy.add_peer(peer_id, best_hash, best_number),
+			SyncingStrategy::StateSyncStrategy(strategy) =>
+				strategy.add_peer(peer_id, best_hash, best_number),
+			SyncingStrategy::ChainSyncStrategy(strategy) =>
+				strategy.add_peer(peer_id, best_hash, best_number),
+		}
+	}
+
+	/// Notify that a peer has disconnected.
+	pub fn remove_peer(&mut self, peer_id: &PeerId) {
+		match self {
+			SyncingStrategy::WarpSyncStrategy(strategy) => strategy.remove_peer(peer_id),
+			SyncingStrategy::StateSyncStrategy(strategy) => strategy.remove_peer(peer_id),
+			SyncingStrategy::ChainSyncStrategy(strategy) => strategy.remove_peer(peer_id),
+		}
+	}
+
+	/// Submit a validated block announcement.
+	///
+	/// Returns new best hash & best number of the peer if they are updated.
+	pub fn on_validated_block_announce(
+		&mut self,
+		is_best: bool,
+		peer_id: PeerId,
+		announce: &BlockAnnounce<B::Header>,
+	) -> Option<(B::Hash, NumberFor<B>)> {
+		match self {
+			SyncingStrategy::WarpSyncStrategy(_) =>
+				Some((announce.header.hash(), *announce.header.number())),
+			SyncingStrategy::StateSyncStrategy(_) =>
+				Some((announce.header.hash(), *announce.header.number())),
+			SyncingStrategy::ChainSyncStrategy(strategy) =>
+				strategy.on_validated_block_announce(is_best, peer_id, announce),
+		}
+	}
+
+	/// Configure an explicit fork sync request in case external code has detected that there is a
+	/// stale fork missing.
+	pub fn set_sync_fork_request(
+		&mut self,
+		peers: Vec<PeerId>,
+		hash: &B::Hash,
+		number: NumberFor<B>,
+	) {
+		match self {
+			SyncingStrategy::WarpSyncStrategy(_) => {},
+			SyncingStrategy::StateSyncStrategy(_) => {},
+			SyncingStrategy::ChainSyncStrategy(strategy) =>
+				strategy.set_sync_fork_request(peers, hash, number),
+		}
+	}
+
+	/// Request extra justification.
+	pub fn request_justification(&mut self, hash: &B::Hash, number: NumberFor<B>) {
+		match self {
+			SyncingStrategy::WarpSyncStrategy(_) => {},
+			SyncingStrategy::StateSyncStrategy(_) => {},
+			SyncingStrategy::ChainSyncStrategy(strategy) =>
+				strategy.request_justification(hash, number),
+		}
+	}
+
+	/// Clear extra justification requests.
+	pub fn clear_justification_requests(&mut self) {
+		match self {
+			SyncingStrategy::WarpSyncStrategy(_) => {},
+			SyncingStrategy::StateSyncStrategy(_) => {},
+			SyncingStrategy::ChainSyncStrategy(strategy) => strategy.clear_justification_requests(),
+		}
+	}
+
+	/// Report a justification import (successful or not).
+	pub fn on_justification_import(&mut self, hash: B::Hash, number: NumberFor<B>, success: bool) {
+		match self {
+			SyncingStrategy::WarpSyncStrategy(_) => {},
+			SyncingStrategy::StateSyncStrategy(_) => {},
+			SyncingStrategy::ChainSyncStrategy(strategy) =>
+				strategy.on_justification_import(hash, number, success),
+		}
+	}
+
+	/// Process block response.
+	pub fn on_block_response(
```

Comment:
One thing I forgot to mention was that in a follow-up we should check if these `on_*_response()` could be refactored into `on_response(PeerId, ProtocolName, request: Vec<u8>, response: Vec<u8>)`.

Reply:
Yeah, we have discussed this and there is this `BlockDownloader` that stands in the way. See the the discussion with Subspace people above